### ISSUE TITLE
Segfault calling `with_logger` in a module 

### ISSUE THREAD

**Issue by xvalcarce** [JuliaLang/julia#45618]
## Description

Julia segfault when calling a function, exported by a module, that calls `with_logger(logger)` where `logger` is a `ConsoleLogger` defined in this module. 

## Step to reproduce

Generate a simple package Foo with src/Foo.jl as follow
```julia
module Foo

export test

using Logging

logger = ConsoleLogger(stdout, Logging.Debug)

function test()
    @info "This works."
    with_logger(logger) do
        @info "This crashes."
    end
end

end
```

The function test() will cause a segmentation fault:
```bash
~/Foo$ julia -e "using Pkg; Pkg.activate(\".\"); using Foo; test()"
```
```
  Activating project at `~/Foo`
[ Info: This works.

signal (11): Segmentation fault
in expression starting at none:1
uv_tty_get_winsize at /workspace/srcdir/libuv/src/unix/tty.c:320
displaysize at ./stream.jl:539
jfptr_displaysize_22405.clone_1 at /opt/julia-1.7.3/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2247 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2429
#handle_message#3 at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/Logging/src/ConsoleLogger.jl:124
handle_message at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.7/Logging/src/ConsoleLogger.jl:109
jfptr_handle_message_37826.clone_1 at /opt/julia-1.7.3/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2247 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2429
macro expansion at ./logging.jl:330 [inlined]
#1 at /home/xvalcarce/Foo/src/Foo.jl:12
unknown function (ip: 0x7f67b2657cff)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2247 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2429
with_logstate at ./logging.jl:511
with_logger at ./logging.jl:623
unknown function (ip: 0x7f67b2657ad1)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2247 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2429
test at /home/xvalcarce/Foo/src/Foo.jl:11
unknown function (ip: 0x7f67b265788f)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2247 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2429
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1788 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:126
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:215
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:166 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:587
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:731
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:885
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:830
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:830
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:944
eval at ./boot.jl:373 [inlined]
exec_options at ./client.jl:268
_start at ./client.jl:495
jfptr__start_22567.clone_1 at /opt/julia-1.7.3/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2247 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2429
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1788 [inlined]
true_main at /buildworker/worker/package_linux64/build/src/jlapi.c:559
jl_repl_entrypoint at /buildworker/worker/package_linux64/build/src/jlapi.c:701
main at julia (unknown line)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x400808)
Allocations: 2721 (Pool: 2712; Big: 9); GC: 0
Segmentation fault (core dumped)
```
## Julia `versioninfo()`

```
Julia Version 1.7.3
Commit 742b9abb4d (2022-05-06 12:58 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, broadwell)
```

**Comment by Seelengrab** [comment]
Works for me on a recent-ish master (plus some minor changes of a PR I'm writing, but that's unrelated):

```
julia> Foo.test()
[ Info: This works.
[ Info: This crashes.

julia> versioninfo()
Julia Version 1.9.0-DEV.620
Commit 605440ab62* (2022-06-07 21:44 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: 4 × Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 4 on 4 virtual cores
Environment:
  JULIA_NUM_THREADS = 4

julia> 
```

Probably needs a bisect, for potential backporting. Does it break on LTS too?

**Comment by xvalcarce** [comment]
@Seelengrab Thanks for the confirmation. This does not only occur from the command line, but also from inside Julia's REPL:

```julia
(@v1.7) pkg> activate .
  Activating project at `~/Foo`

julia> using Foo

julia> test()
[ Info: This works.

signal (11): Segmentation fault
```

Note that declaring the module directly in the REPL, i.e. having `Foo` defined in `Main`, works fine
```julia
julia> include("./src/Foo.jl")
Main.Foo

julia> using .Foo

julia> test()
[ Info: This works.
[ Info: This crashes.
```
Could the segfault be due to `logger` being created during the precompilation of `Foo`? `stdout` (the stream argument) might not be a valid stream outside of the precompilation time.

**Comment by xvalcarce** [comment]
Found the cause!
The `stream` used in the ConsoleLogger is `invalid status`:

```julia
(@v1.7) pkg> activate .
  Activating project at `~/Foo`

julia> using Foo

julia> Foo.logger.stream
Base.TTY(RawFD(4294967295) invalid status, 0 bytes waiting)

julia> Foo.logger.stream.handle
Ptr{Nothing} @0x0000000000000000

julia> Foo.logger.stream.status
3

julia> isopen(Foo.logger.stream)
true
```
the `invalid status` string comes from the fact that `stream.handle` is the null pointer `C_NULL` and that `stream.status` is 3 (`StatusOpen`). 
When logging, the function `handle_message` which sends the message to the stream check if the stream is open

https://github.com/JuliaLang/julia/blob/7dc132cf5a44717dbb619812dc05fd127c9d805a/stdlib/Logging/src/ConsoleLogger.jl#L120-L123

but `isopen(stream)` does not check for `stream.handle` in Julia v1.7.3 (nor in LTS). 
However this is fixed by https://github.com/JuliaLang/julia/commit/72cd1c454cd30c5b9bd55611add0576d7c55878f# , specifically in `base/stream.jl`:

```diff
- if x.status == StatusUninit || x.status == StatusInit
+ if x.status == StatusUninit || x.status == StatusInit || x.handle === C_NULL
```

**Comment by giordano** [comment]
> However this is fixed by 72cd1c4

Meaning that this issue can be closed?

**Comment by KristofferC** [comment]
That part could be backported.
